lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,136 @@
+    HasAllowRole(user, login, node, role) <- HasRole(user, role), HasAllowNodeLabel(role, node, _, _), RoleAllowsLogin(role, login),
+        !RoleDeniesLogin(role, login);
+    HasAllowRole(user, login, node, role) <- HasRole(user, role), HasAllowNodeLabel(role, node, _, _), HasTrait(user, LOGIN_TRAIT_HASH, login),
+        !RoleDeniesLogin(role, login), !RoleDeniesLogin(role, LOGIN_TRAIT_HASH);
+    HasDenyRole(user, node, role) <- HasRole(user, role), HasDenyNodeLabel(role, node, _, _);
+    HasDeniedLogin(user, login, role) <- HasRole(user, role), RoleDeniesLogin(role, login);
+    HasDeniedLogin(user, login, role) <- HasRole(user, role), HasTrait(user, LOGIN_TRAIT_HASH, login), RoleDeniesLogin(role, LOGIN_TRAIT_HASH);
+
+    // HasAccess rule determines each access for a specified user, login and node
+    HasAccess(user, login, node, role) <- HasAllowRole(user, login, node, role), !HasDenyRole(user, node, _), !HasDeniedLogin(user, login, _);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasTrait(user, LOGIN_TRAIT_HASH, login);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasAllowRole(user, login, node, _);
+
+    DenyLogins(user, login, role) <- HasDeniedLogin(user, login, role);
+}
+
+#[no_mangle]
+pub extern "C" fn process_access(input: *mut c_uchar, output: *mut c_uchar, input_len: size_t, output_len: size_t) -> i32 {
+    let mut runtime = Crepe::new();
+    let b = unsafe { slice::from_raw_parts_mut(input, input_len) };
+    let r = match types::Facts::decode(BytesMut::from(&b[..])) {
+        Ok(b) => b,
+        Err(_e) => return Errors::InvalidInputError as i32
+    };
+
+    for pred in &r.predicates {
+        if pred.name == types::facts::PredicateType::Hasrole as i32 {
+            runtime.extend(&[HasRole(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Hastrait as i32 {
+            runtime.extend(&[HasTrait(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Roleallowslogin as i32 {
+            runtime.extend(&[RoleAllowsLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Roledenieslogin as i32 {
+            runtime.extend(&[RoleDeniesLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Roleallowsnodelabel as i32 {
+            runtime.extend(&[RoleAllowsNodeLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Roledeniesnodelabel as i32 {
+            runtime.extend(&[RoleDeniesNodeLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Nodehaslabel as i32 {
+            runtime.extend(&[NodeHasLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else {
+            return Errors::InvalidInputError as i32;
+        }
+    }
```

Comment:
Is this unchecked indexing safe? i.e can you prove that `pred.atoms` is long enough. Otherwise you need to check the length to avoid UB on a panic.

Reply:
The atoms of the predicate should be safe, since we manually create each one on the Go side so that it matches the specified length on the Rust end.